### PR TITLE
Ruby on Rails capstone project - Budget app

### PR DESCRIPTION
**Pull request by itsmraga-hub**
# 💯 Achievements

- Use ruby gems as software packages system.
- Install Ruby on Rails framework.
- Understand Rails RESTful design and router.
- Use controllers to handle requests and render empty views.
- Use params from browser request in a safe way.
- Use preprocessed html file with embedded Ruby code.
- Use layouts and templates for shared content.
- Use database migration files to maintain database schema.
- Use validations for models.
- Secure app from n+1 problems.
- Understand what ORM is.
- Write SQL queries with ActiveRecord.
- Set up associations between models.
- Build a webapp that requires the user to log in.
- Use devise gem for authentication.
- Limit access to webapp resources based on authorization rules.
- Analyze in writing why you have made a coding choice using one structure over another.

Tests

Screenshot (130)

Splash screen

image

Categories page

image

Details page

image

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+<!-- LIVE DEMO -->
+
+## 🚀 Live Demo <a name="live-demo"></a>
+
+> Add a link to your deployed project.
+
+- [Live Demo Link coming soon](https://yourdeployedapplicationlink.com)
```
Comment (julie-ify) [inline README.md]: - [x] Kindly deploy your app and include its link in the readme file.

- [x] Kindly record a video presentation of your project and include the link in the readme file

File: app/views/expenses/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    <% @expenses.each do |expense| %>
+      <%= render expense %>
+    <% end %>
+  </div>
+  <div class="total-container">
+  <p>TOTAL PAYMENT</p>
+  <p class="total">
+      $<%= @budget_category.total%>
+  </p>
+  </div>
```
Comment (julie-ify) [inline app/views/expenses/index.html.erb:31]: - [ ] The total payment should be on the top of the page, not the bottom. Kindly fix that.

File: app/views/expenses/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<section class="devise-forms">
+  <h1>New expense</h1>
+
+  <%= render "form", expense: @expense %>
+
+  <br>
+
+  <div>
+    <%= link_to 'Back to Budget Categories', budget_categories_path %>
```
Comment (julie-ify) [inline app/views/expenses/new.html.erb]: - [ ] This back button should be **`<`** and should be on the top of the page based on the design.

- [ ] When I am on this new expense page and I click the back button I expect to be taken to the expense index page, not the categories page. Kindly fix that.

File: app/views/budget_categories/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<section class="devise-forms">
+  <h1>New budget category</h1>
+
+  <%= render "form", budget_category: @budget_category %>
+
+  <br>
+
+  <div>
+    <%= link_to "Back to budget categories", budget_categories_path %>
```
Comment (julie-ify) [inline app/views/budget_categories/new.html.erb]: - [ ] This back button should be **`<`** and should be on the top of the page based on the design.

File: app/views/expenses/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,37 @@
+<section>
+  <p style="color: green"><%= notice %></p>
+
+  <div class="header header-details">
+    <%= link_to budget_categories_path do %>
+      <%# <i class="fa-solid fa-arrow-left-long"></i> %>
```
Comment (julie-ify) [inline app/views/expenses/index.html.erb]: - [x] Kindly remove commented code.

**Review by julie-ify — CHANGES_REQUESTED** [review]
## Hi @itsmraga-hub,

Good job so far!

## Highlights!
- Good feature branch name ✔️ 
- No linter errors ✔️ 
- Descriptive pull request ✔️ 
- Descriptive commit messages ✔️ 
- Good code organization and working app ✔️
- Descriptive README ✔️ 

Hello ✋, you have done well reaching this point of the milestone 🤝 . But there are some issues that you still need to work on to go to the next project but you are almost there!

# Status: Required Changes ♻️⛔️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread and tag me `@julie-ify` if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: README.md (hunk 1 above)
Comment (itsmraga-hub) [reply README.md]: The video link is on PR, but I;ll add it to README

File: app/views/expenses/index.html.erb (hunk 2 above)
Comment (itsmraga-hub) [reply app/views/expenses/index.html.erb:31]: I get this but in the design the total is at the bottom of the page, after the list of transactions @julie-ify

File: app/views/expenses/index.html.erb (hunk 5 above)
Comment (itsmraga-hub) [reply app/views/expenses/index.html.erb]: Done

File: app/views/expenses/new.html.erb (hunk 3 above)
Comment (itsmraga-hub) [reply app/views/expenses/new.html.erb]: Done. Thanks.

File: app/views/budget_categories/new.html.erb (hunk 4 above)
Comment (itsmraga-hub) [reply app/views/budget_categories/new.html.erb]: Done. Thanks

**Review by Reem-lab — CHANGES_REQUESTED** [review]
..

File: app/views/expenses/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<section class="devise-forms">
+<div class="header header-details">
+    <%= link_to budget_categories_path do %>
```
Comment (Reem-lab) [inline app/views/expenses/new.html.erb]: - [ ] As requested by the privous code reviewer when you are in the page for creating a new expense the back arrow should return to the expenses page not the category so kindly change the path here 🙏🏻✅✨💯

## to the transaction page 💯💯💯
image


- Your design


https://user-images.githubusercontent.com/58553711/208921731-7732bf2a-07cf-4fe0-b3b3-10f2e53ae590.mp4

**Review by Reem-lab — CHANGES_REQUESTED** [review]
# Hi @itsmraga-hub, 👋🏻👋🏻 

Great work on making the changes requested by a previous reviewer 👏 💯✨✅
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights ✨✨
- No Linters errors ✅ 
- Followed Gitflow ✅
- Great working on the functionality🔥🚀🎉
- Professional README ✅
- All testing passed ✅
- Descriptive PR ✅
- Good working on the deployement ✅

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/devise/shared/_error_messages.html.erb [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+                 resource: resource.class.model_name.human.downcase)
+       %>
+    </h2>
+    <ul>
+      <% resource.errors.full_messages.each do |message| %>
+        <li><%= message %></li>
+      <% end %>
+    </ul>
+  </div>
+<% end %>
```
Comment (Reem-lab) [inline app/views/devise/shared/_error_messages.html.erb]: - [ OPTIONAL ] I would highly recommend to add some styles to your error message to make your project has more human sense and that's important for user experience ✅✨💯🚀

**Review by Reem-lab — CHANGES_REQUESTED** [review]

File: app/views/devise/shared/_error_messages.html.erb (hunk 7 above)
Comment (itsmraga-hub) [reply app/views/devise/shared/_error_messages.html.erb]: Done

File: app/views/expenses/new.html.erb (hunk 6 above)
Comment (itsmraga-hub) [reply app/views/expenses/new.html.erb]: Fixed.

**Review by CollinsTatang — APPROVED** [review]
# Approved 👏  
### **Highlights 🚀 :heavy_check_mark: :heavy_check_mark:** 

- The repo looks professional.:heavy_check_mark:
- Git/GitHub flow best practices have been followed:heavy_check_mark:
- You have descriptive PR:heavy_check_mark:

In my opinion, your project is ready for presentation!  There is nothing else to say other than... it's time to merge it :shipit: 
### **Congratulations!** 🎉 
image



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @CollinsTatang in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by itsmraga-hub** [comment]
Thanks @CollinsTatang, @Reem-lab and @julie-ify. Cheers to y'all.

**Pull request merged**